IdentityCore/src/cache/mac/MSIDMacKeychainTokenCache.m

```diff
@@ -870,6 +973,121 @@ - (BOOL)clearWithContext:(id<MSIDRequestContext>)context
+        
+        return self.accessForNonSharedBlob;
+    }
+}
+
+- (NSArray *)createAccessForCurrentApp:(id<MSIDRequestContext>)context error:(__unused NSError **)error
+{
+    SecTrustedApplicationRef trustedApplication = nil;
+    OSStatus status = SecTrustedApplicationCreateFromPath(nil, &trustedApplication);
+    if (status != errSecSuccess)
+    {
+        [self createError:@"Failed to create trusted application for current application path"
+                   domain:MSIDKeychainErrorDomain errorCode:status error:error context:context];
+        
+        return nil;
+    }
+    
+    NSArray *trustedApplications = @[(__bridge_transfer id)trustedApplication];
+    return trustedApplications;
+}
+
+- (id)accessCreateWithChangeACL:(NSArray<id> *)trustedApplications
+                         context:(id<MSIDRequestContext>)context
+                           error:(NSError **)error
+{
+    SecAccessRef access;
+    OSStatus status = SecAccessCreate((__bridge CFStringRef)s_defaultKeychainLabel, (__bridge CFArrayRef)trustedApplications, &access);
+    
+    if (status != errSecSuccess)
+    {
+        [self createError:@"Failed to create SecAccessRef object for blob."
+                   domain:MSIDKeychainErrorDomain errorCode:status error:error context:context];
+        return nil;
+    }
+    
+    if (![self accessSetACLTrustedApplications:access
+                           aclAuthorizationTag:kSecACLAuthorizationChangeACL
+                           trustedApplications:trustedApplications
+                                       context:context
+                                         error:error])
+    {
+        [self createError:@"Failed to update ACL with kSecACLAuthorizationChangeACL authorization tag and set trusted applications."
+                   domain:MSIDKeychainErrorDomain errorCode:status error:error context:context];
+        
+        CFRelease(access);
+        return nil;
+    }
+    
+    return CFBridgingRelease(access);
+}
+
+- (BOOL)accessSetACLTrustedApplications:(SecAccessRef)access
+                     aclAuthorizationTag:(CFStringRef)aclAuthorizationTag
+                     trustedApplications:(NSArray<id> *)trustedApplications
+                                 context:(id<MSIDRequestContext>)context
+                                   error:(NSError **)error
+{
+    OSStatus status;
+    NSArray *acls = (__bridge_transfer NSArray*)SecAccessCopyMatchingACLList(access, aclAuthorizationTag);
+    // TODO: handle case where tag is not found?
+    for (id acl in acls)
+    {
+        CFStringRef description;
+        CFArrayRef oldtrustedAppList;
+        SecKeychainPromptSelector selector;
+        
+        status = SecACLCopyContents((__bridge SecACLRef)acl, &oldtrustedAppList, &description, &selector);
+        
+        if (status != errSecSuccess)
+        {
+            [self createError:@"Failed to get contents from ACL" domain:MSIDKeychainErrorDomain errorCode:status error:error context:context];
```

Comment:
what are some possible reasons for failures and how recoverable they should be? This right now is completely unrecoverable for MSAl.

Reply:
I think only possible errors are internal keychain error because every SecAccessRef must have a corresponding ACL list which internally should all have a description, authorization tags and list of trusted applications and this api is trying to extract that acl list.